.github/workflows/test.yml

```diff
@@ -341,7 +341,12 @@ jobs:
       with:
         template: templates/vmnet.yaml
     - name: Install test dependencies
-      run: brew install qemu bash coreutils iperf3
+      run: |
+        brew install bash coreutils iperf3
+        # Install qemu 8.2.1. Unfortunately this mechanism builds from source.
+        brew tap-new lima/qemu
+        brew extract --version=8.2.1 qemu lima/qemu
+        brew install lima/qemu/qemu@8.2.1
```

Comment:
There is a precompiled binary that we should be able to download, but I couldn't get it to work, and am out of time. I will try this again tomorrow; for now I'm just hoping that this will fix the test. We can optimize it afterwards, once it is confirmed that it works.

We probably want to switch to fetch precompiled binaries (at the current pinned version) for the other dependencies too, as brew will start compiling from source once the formulas get updated because they will not build updated bottles.

Reply:
`brew extract` removes bottle information. If you download the formula directly via `curl` from the commit where the bottle was updated, you should be able to install it using the bottle.
e.g.
`curl https://raw.githubusercontent.com/Homebrew/homebrew-core/4c7ffca0dc9ced8cd25d1019c49f22e2cb3c8b24/Formula/q/qemu.rb -o $(brew --repo lima/qemu)/Formula/qemu@8.2.1.rb`

Reply:
It is not so easy; the formula also needs to change the class name when it is exported. I wrote a function that does all this:

```shell
#!/bin/bash -eux -o pipefail

# This function only works for formulas in the homebrew-core
brew_install_version() {
    local FORMULA=$1
    local VERSION=$2

    TAP=lima/qemu

    # Extract formula to local tap (this will strip the bottle info)
    EXTRACT=$(brew extract --verbose --force --version "$VERSION" "$FORMULA" "$TAP")

    # Get the commit id for this version
    SHA=$(perl -ne 'print $1 if /from revision ([0-9a-f]+) to/' <<<"$EXTRACT")

    # File path of the extracted formula
    OUTPUT="$(brew --repo "$TAP")/Formula/${FORMULA}@${VERSION}.rb"

    # Get versioned class name of extracted formula
    AT=$(perl -ne 'print $1 if /^class (\w+AT\w+) < Formula/' "$(brew formula "$OUTPUT")")

    # Find the location of the formula inside the homebrew-core, 'Formula/?/$FORMULA'
    FILE=$(brew formula "$FORMULA" | sed 's#^.*/homebrew-core/##')

    REPO=$(brew --repo homebrew/homebrew-core)

    # Overwrite the formula with the original one including bottle info
    git -C "$REPO" show "$SHA:$FILE" >"$OUTPUT"

    # Replace class name with versioned class name
    perl -p -i -e "s/class \w+ < Formula/class $AT < Formula/" "$OUTPUT"
    echo "brew install ${TAP}/qemu@${VERSION}"
}

brew_install_version qemu 8.2.1
```

Unfortunately it still fails because brew doesn't add authentication to the bottle download requests:

```
==> Fetching lima/qemu/qemu@8.2.1
==> Downloading https://ghcr.io/v2/homebrew/core/qemu/8.2.1/blobs/sha256:41c77f6bac3e8c1664c665fbe2b19b19ee9da57f8e1ad6697348286710cc3575
curl: (56) The requested URL returned error: 404
```

I gave up and just downloaded the bottle directly with `curl`.

Reply:
It's better to extract directly from the repository instead of using `brew extract`. Instead of using the commit that `brew extract` would use, find the one with the commit message 'qemu: update 8.2.1 bottle.' This way, you'll get a formula with a valid bottle.

The following is a script that reflects those changes.
```bash
#!/bin/bash -eux -o pipefail

# This function only works for formulas in the homebrew-core
brew_install_version() {
    local FORMULA=$1
    local VERSION=$2

    TAP=lima/qemu

    HOMEBERW_CORE_TAP=homebrew/homebrew-core
    REPO=$(brew --repo "${HOMEBERW_CORE_TAP}")

    # Get the commit id for this version to update the bottle info
    SHA=$(git -C "${REPO}" log -1 :/"${FORMULA}: update ${VERSION} bottle" | awk '/^commit/{print $2}')

    # File path of the extracted formula
    OUTPUT="$(brew --repo "$TAP")/Formula/${FORMULA}.rb"

    # Find the location of the formula inside the homebrew-core, 'Formula/?/$FORMULA'
    FILE=$(brew formula "${FORMULA}" | sed "s#^.*/${HOMEBERW_CORE_TAP}/##")


    # Extract the formula file from the homebrew-core repository
    git -C "$REPO" show "$SHA:$FILE" >"$OUTPUT"

    echo "brew install ${TAP}/qemu"
}

brew_install_version qemu 8.2.1
```

Reply:
I don't want to make any more changes; the hard-coded URL should work fine.

I've spent a day and a half on this issue, and I don't want to spend any more (with the exception that I will try to run Monterey in a nested VM, to see if this causes a problem with QEMU).

I would like to see this PR merged, so we can run the upgrade test again, instead of ignoring it because it is currently always being skipped.

Reply:
@norio-nomura Thank you for fixing the script though. 🙏

I'll keep it around because it is more generally useful in other situations as well.

Reply:
> I will try to run Monterey in a nested VM, to see if this causes a problem with QEMU

It did seem to freeze the VM the first time I started a Lima instance inside the VM, but in subsequent tests it eventually started. CPU usage got very high for a minute or two, so it is still conceivable that this is responsible for dropping the connection to the runner.

The modified script didn't work on the VM for some reason, but I'm tired out of dealing with brew, so I will give it a rest for a while.

Reply:
The newly installed Homebrew should not have `homebrew/core` tapped, so you need to either tap it or modify the script to retrieve Formula information from the remote.

Reply:
Obtaining the `sha` looks like this:
```console
$ gh search commits --repo homebrew/homebrew-core "qemu: update 8.2.1 bottle" --json sha --jq "last|.sha"
4c7ffca0dc9ced8cd25d1019c49f22e2cb3c8b24
```

Reply:
Looks like I forgot the `--force` option; it refuses to tap it unless you add `--force`: `brew tap --force homebrew/core`. But I can see why they want to discourage users from tapping it.

Reply:
> Obtaining the `sha` looks like this:

Oh, I like that one, thanks! 🙏